ADBDEV-697. Fix null transition for non-strict aggregate functions

dursegov opened the pull request:
The bug was indeed related to the case when pergroupstate->transValueIsNull is false, but pergroupstate->transValue = 0.
These fields correspond to an aggregate group and updated on HashAggTable insert:
agg_hash_initial_pass -> advance_aggregates -> advance_transition_function -> invoke_agg_trans_func

As can be seen new transValue is returned from int8_avg_combine(fcinfo) and transValueIsNull is set to fcinfo->isnull that is set to false by default. The problem that int8_avg_combinei() can return immediately the old transValue and not touch fcinfo->isnull at all if increment is null. So we can end up with transValueIsNull is false and transValue = 0. This is possible, because int8_avg_combine() is defined as strict function (transfn->fn_strict = true) and accept nulls. As another approach it is good to check if this flag is really required for this function.

darthunix reviewed (changes requested):
I think the root of all problems is corrupted Datums in some aggregation state hash entries (no matter whether they were spilled or not).  So the goals of this PR should be:
1. Add asserting elogs in places where it can cause a segfault. We should not hide a problem, or we get wrong results in aggregation.
2. Try to find out the reason of this corruption.
At the moment this fix seems to be wrong (but I still have no idea about a place where aggregation state was corrupted)

Commit:
Fix pergroupstate for non-strict transition functions

On some datasets GP fails with segfault on serialization of a
HHashAgg table entries to spillfiles. This problem is caused by a
pergroupstate inconsistency in some entries: transValue is 0 when
transValueIsNull and noTransValue are false.

The reason of this inconsistency is a wrong fcinfo->isnull
status returned by some of non-strict *_combine transition
functions. The first part of the problem is that *_combine functions
do not modify fcinfo->isnull state but can return NULL. The second
part of it is that fcinfo->isnull in invoke_agg_trans_func() is always
set false by default for non-strict transition functions. As a result
pergroupstate can become inconsistent and cause segfailts.

We can get information whether a transition function returns NULL
or not only from its internals. So this commit fixes *_combine
functions to set fcinfo->isnull correctly.

Co-authored-by: Dmitry Ursegov <uda@arenadata.io>

darthunix commented:
I have compared results on the same data set for this patch on GP and on upstream PG - they return the same results.

deart2k commented:
bender build
